content/r-package-dev/Writing_Tests.Rmd

```diff
@@ -0,0 +1,153 @@
+
+Applying these techniques where applicable will develop a solid suite of tests that will instill confidence in those developing on or against your package.  It is helpful to gather code coverage metrics to quantify the level of testing that exists.  The R package `covr` is available to generate these metrics and push them to a location to share the results.
+
+## How do I write tests?
+
+We are going to build on what was covered in the documentation section by writing some tests.  Here is a reminder of the pH validity function we created earlier.
+
+```{r ph-setup, include=TRUE, warning=FALSE, message=FALSE}
+is.valid.pH <- function(pH){
+  pH >= 0 & pH <= 14
+}
+```
+
+```{r reporter-setup, include=FALSE, warning=FALSE, message=FALSE}
+  reporter <- testthat:::find_reporter("list")
+  old <- testthat:::set_reporter(reporter)
+```
+
+The `testthat` package provides the functions we are going to use for testing.  Files performing the testing should be placed in the `tests/testthat` directory and a `testhat.R` script should invoke `test_check()` on your package.
+
+```{r ph-valid, include=TRUE, warning=FALSE, message=FALSE}
+library(testthat)
+context("Valid pH values")
+test_that("pH values inside valid range return true", {
+  expect_true(is.valid.pH(7))
+})
+
+```  
+
+The `test_that()` is provided the test name in BDD form.  The second argment to this function is the code block containing the code to exercise the functionality and expectations of the results.  The `expect_*()` calls are used to declare these expectations, and will fail when the expectation is not met.  You'll also note that `context()` is called before running the test case, this groups a number of test cases together into logical chunks to add a bit more information to the test reporting.
+
+We also want to make sure that non-valid pH values are correctly identified.
+```{r ph-invalid, include=TRUE, warning=FALSE, message=FALSE}
+test_that("pH values outside valid range return false", {
+  expect_false(is.valid.pH(-3))
+  expect_false(is.valid.pH(15))
+})
+```
+
+We are able to make several assertions in one test, but this should not be overused as we want the assertions to match the BDD description as much as possible.  Lastly, we are going to add edge cases to this test to verify that they are correct
+```{r ph-edges, include=TRUE, warning=FALSE, message=FALSE}
+test_that("pH edge cases return true", {
+  expect_true(is.valid.pH(0))
+  expect_true(is.valid.pH(14))
+})
+```
+
+Through a bit of magic we can look at the results of these tests.  Normally this will happen by running "Build > Test Package" through RStudio (Ctrl+Shift+T shortcut).
+
+```{r test-output, include=TRUE, warning=FALSE, message=FALSE}
+  reporter$get_results() # only works for ListReporter
```

Comment:
i like that the other `reporter` blocks are hidden and that you've noted the Ctrl+Shift+T option in the text. This particular code line shows up in the md. Can you hide it and just show the output with `echo=FALSE`?